### PR TITLE
Base of Solvency calculation for fpna

### LINKED ISSUE #133: Minimization of primaryScope during the BE Reimport
Lets assume this use case:
1) Import BE
2) Import Scenario X
3) Reimport of one GIC1 for BE. 

The last import computes GIC1 for BE and all GICs in Scenario. For Scenario, instead, it should recompute only GIC1.

### PR DESCRIPTION
**Pull request by dcolleoni**
fpna: 
- design the structure of Storage (generic and for pillars) and scopes
- importer for SolvencyCashflow
- add "QueryCashflowAsync"
- add basic tests

- modify data model for Cashflows
- enhance Solvency II AmountType hierarchy
- add Risk Driver
- correct & modify extensions Methods

ifrs17 : 
improve the QueryPartitionedDataAsync to allow for filter [#133 ]

### CODE REVIEW

File: fpna-template/Initialization/InitSystemorphToMemory.ipynb [hunk 1]
```diff
@@ -38,8 +66,7 @@
         {
             "cell_type": "code",
             "source": [
-                "await DataSource.SetAsync();",
-                "\nDataSource.Reset(x => x.ResetCurrentPartitions());"
+                "#!import \"../../ifrs17-template/Initialization/InitSystemorphBaseToMemory\""
```
Comment (dcolleoni) [inline fpna-template/Initialization/InitSystemorphToMemory.ipynb]: fix the path

File: ifrs17/Import/Importers.ipynb [hunk 2]
```diff
@@ -508,7 +508,8 @@
                 "\n                            Year = args.Year, ",
                 "\n                            Month = DefaultDataNodeActivationMonth, ",
                 "\n                            State = State.Active, ",
-                "\n                            Partition = partition.Id})",
+                "\n                            Partition = partition.Id",
+                "\n                            Scenario = args.Scenario})",
```
Comment (dcolleoni) [inline ifrs17/Import/Importers.ipynb:515]: Run tests for these changes

File: fpna/DataModel/DataStructure.ipynb [hunk 3]
```diff
@@ -192,23 +276,32 @@
         {
             "cell_type": "code",
             "source": [
-                "public record ReferencedCashflow : SimpleCashflow",
+                "//Remove ",
```
Comment (dcolleoni) [inline fpna/DataModel/DataStructure.ipynb]: Clean up code

File: fpna/DataModel/DataStructure.ipynb [hunk 4]
```diff
@@ -217,6 +310,38 @@
...
+                "## Solvency Variable"
+            ],
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "code",
+            "source": [
+                "//Remove",
```
Comment (dcolleoni) [inline fpna/DataModel/DataStructure.ipynb]: Remove

File: fpna/DataModel/DataStructure.ipynb [hunk 5]
```diff
@@ -249,6 +375,29 @@
             "metadata": {},
             "execution_count": 0,
             "outputs": []
+        },
+        {
+            "cell_type": "markdown",
+            "source": [
+                "## Solvency Report Variable"
```
Comment (dcolleoni) [inline fpna/DataModel/DataStructure.ipynb]: Remove

File: fpna/DataModel/DataStructure.ipynb [hunk 6]
```diff
@@ -46,6 +46,77 @@
...
+                "\n}"
+            ],
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "code",
+            "source": [
+                "public interface IWithRiskDriver",
```
Comment (dcolleoni) [inline fpna/DataModel/DataStructure.ipynb]: Is this needed?

File: fpna-template/Files/Cashflows.ipynb [hunk 7]
```diff
@@ -50,9 +50,23 @@
...
+                "\n\";"
+            ],
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "code",
+            "source": [
+                "var solvencyCashflows = @\"@@SolvencyCashflow",
```
Comment (dcolleoni) [inline fpna-template/Files/Cashflows.ipynb]: Remove

File: fpna/DataModel/DataStructure.ipynb [hunk 8]
```diff
@@ -90,49 +165,31 @@
...
-                "\n{",
-                "\n    public double InitialValue { get; init; }",
-                "\n",
-                "\n    public int Length { get; init; }",
-                "\n",
-                "\n    public int Shift { get; init; }",
+                "\n    [IdentityProperty]//[Dimension(typeof(Pillar))] ",
+                "\n    public string Pillar { get; init; }",
                 "\n",
+                "\n    //[Dimension(typeof(double), nameof(Values))]",
```
Comment (dcolleoni) [inline fpna/DataModel/DataStructure.ipynb:190]: remove

File: fpna/DataModel/DataStructure.ipynb [hunk 9]
```diff
@@ -151,7 +208,12 @@
         {
             "cell_type": "code",
             "source": [
-                "public record BasicCashflow : SimpleCashflow {}"
+                "public record BasicCashflow : Cashflow, IBased",
+                "\n{",
+                "\n    public double InitialValue { get; init; }",
```
Comment (dcolleoni) [inline fpna/DataModel/DataStructure.ipynb:214]: InitialValue, Length, Shift and Shape belongs to the same type as they are all related on computation. 
I would move Shape from Cashflow to IBased.
@sbatka what do you think?

File: fpna/DataModel/DataStructure.ipynb [hunk 10]
```diff
@@ -192,7 +256,7 @@
         {
             "cell_type": "code",
             "source": [
-                "public record ReferencedCashflow : SimpleCashflow",
+                "public record ReferencedCashflow : Cashflow, IReferenced",
```
Comment (dcolleoni) [inline fpna/DataModel/DataStructure.ipynb]: ReferencedCashflow should also inherit from IBased.
@sbatka should it also inherit from IBounded?

File: fpna/Import/Scopes.ipynb [hunk 11]
```diff
@@ -72,10 +72,23 @@
...
+                "\n}"
+            ],
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "code",
+            "source": [
+                "public interface BasicCashflows2 : IScope<CashflowIdentity, ICashflowStorage>",
```
Comment (dcolleoni) [inline fpna/Import/Scopes.ipynb]: remove

File: fpna/Import/Storage.ipynb [hunk 12]
```diff
@@ -50,76 +82,255 @@
...
+                "# Class"
+            ],
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "code",
+            "source": [
+                "// public class CashflowStorage",
```
Comment (dcolleoni) [inline fpna/Import/Storage.ipynb]: Clean up

File: ifrs17/Utils/Queries.ipynb [hunk 13]
```diff
@@ -450,23 +450,46 @@
...
+                "\n}"
+            ],
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "code",
+            "source": [
+                "public static async Task<T[]> QueryPartitionedDataAsync<T,P>(this IWorkspace workspace, IDataSource dataSource, ",
```
Comment (dcolleoni) [inline ifrs17/Utils/Queries.ipynb:445]: clean up

File: ifrs17/Utils/Queries.ipynb [hunk 14]
```diff
@@ -450,23 +450,46 @@
...
+                "\n                                                                  Expression<Func<T, bool>> filterExpression = null)",
                 "\n    where T : IPartitioned",
                 "\n    where P : IPartition",
                 "\n{",
-                "\n    var isRelaxed = ((format != ImportFormats.Cashflow && typeof(T).Name == nameof(IfrsVariable)) ||",
-                "\n                    (format == ImportFormats.Cashflow && typeof(T).Name == nameof(RawVariable)));",
+                "\n    //If Ifrs17 format",
+                "\n    var isRelaxed = GetIsRelaxed<T>(format);",
                 "\n",
-                "\n    var variablesFromWorkspace = await workspace.LoadPartitionedDataAsync<T,P>(targetPartition);",
```
Comment (sbatka) [inline ifrs17/Utils/Queries.ipynb:452]: LoadPartitionedDataAsync this method throws an error. (No overload for method 'LoadPartitionedDataAsync' takes 2 arguments).

File: fpna/DataModel/DataStructure.ipynb (hunk 10 above)
Comment (sbatka) [reply fpna/DataModel/DataStructure.ipynb]: yes agreed

File: fpna/DataModel/DataStructure.ipynb (hunk 9 above)
Comment (sbatka) [reply fpna/DataModel/DataStructure.ipynb:214]: agreed

**Review by sbatka — APPROVED** [review]

**Pull request merged**